base/REPLCompletions.jl

```diff
@@ -381,12 +381,44 @@ function bslash_completions(string, pos)
     return (false, (String[], 0:-1, false))
 end
 
+function dict_identifier_key(str,tag)
+    if tag === :string
+        str_close = str*"\""
+    elseif tag === :cmd
+        str_close = str*"`"
+    else
+        str_close = str
+    end
+
+    frange, end_of_indentifier = find_start_brace(str_close, c_start='[', c_end=']')
+    isempty(frange) && return (nothing, nothing, nothing)
+    identifier = Symbol(str[frange[1]:end_of_indentifier])
+    isdefined(Main,identifier) || return (nothing, nothing, nothing)
+    begin_of_key = findnext(x->!in(x,whitespace_chars), str, end_of_indentifier+2)
+    partial_key = str[begin_of_key:end]
+    typeof(eval(Main,identifier)) <: Associative || return (nothing, nothing, nothing)
+    identifier, partial_key, begin_of_key
+end
+
 function completions(string, pos)
     # First parse everything up to the current position
     partial = string[1:pos]
     inc_tag = Base.syntax_deprecation_warnings(false) do
         Base.incomplete_tag(parse(partial, raise=false))
     end
+
+    # if completing a key in a Dict
+    identifier, partial_key, loc = dict_identifier_key(partial,inc_tag)
+    if identifier != nothing
+        matches = []
+        for key in eval(Main,:(collect(keys($identifier))))
+            startswith(repr(key),partial_key) && push!(matches,repr(key))
```

Comment:
I think that we have generally tried to stay away from repl features that
can execute arbitrary code while you are editing. Could we avoid that by
supporting just a reasonable subset of key types?
On 20 Jun 2016 09:36, "Ivar Nesje" <notifications@github.com> wrote:

> In base/REPLCompletions.jl
> <https://github.com/JuliaLang/julia/pull/17017#discussion_r67646639>:
>
> >  function completions(string, pos)
> >      # First parse everything up to the current position
> >      partial = string[1:pos]
> >      inc_tag = Base.syntax_deprecation_warnings(false) do
> >          Base.incomplete_tag(parse(partial, raise=false))
> >      end
> > +
> > +    # if completing a key in a Dict
> > +    identifier, partial_key, loc = dict_identifier_key(partial,inc_tag)
> > +    if identifier != nothing
> > +        matches = []
> > +        for key in eval(Main,:(collect(keys($identifier))))
> > +            startswith(repr(key),partial_key) && push!(matches,repr(key))
>
> Is the call to collect really needed, and should we use a temporary
> variable to avoid calling repr(key) twice?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaLang/julia/pull/17017/files/17da9f809389e84c028c0404833323031740e7be#r67646639>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABjTf9KSHjxCXeSCCqlCKunOGuGnIK6Rks5qNkMMgaJpZM4I5Ln->
> .
>